lib/cwrap.c

```diff
@@ -0,0 +1,245 @@
+ */
+#include <limits.h>
+#include <stdlib.h>
+#include <stdio.h>
+#include <unistd.h>
+#include <errno.h>
+#include <string.h>
+#include"../include/cwrapHeader.h"
+#include <limits.h>
+#include <sys/types.h>
+#include <dirent.h>
+#include <sys/stat.h>
+#include <fcntl.h>
+#include<assert.h>
+
+
+static Map* map;
+
+Map *initializeMap(int capacity){
+	int i;
+	Map *map=(Map*)malloc(sizeof(Map));
+
+	map->opened_files=(opened_dir_struct*)malloc(capacity*sizeof(opened_dir_struct));
+	assert(map->opened_files!=NULL);
+	map->capacity=capacity;
+	map->length=0;
+	return map;
+
+}
+
+Map* getMap(){
+	if(map==0){
+		map=initializeMap(4);
+	}
+	return map;
+}
+//split file from path
+char* split_path_file(char *relative_path) {
+    const char slash='/';
+   	char *filename;
+	char *dirName;
+	filename= strrchr(relative_path, slash);
+	dirName=strndup(relative_path,strlen(relative_path)- strlen(filename));
+	return dirName;
+}
+//check the capacity of map
+int checkCapacity(){
+	if(map->length>=map->capacity)
+		return -1;
+	else
+		return 0;
+}
+// increases the capacity of map
+Map* increaseMapCapacity(){
+	int i;opened_dir_struct *new_opened_files;
+	new_opened_files=(opened_dir_struct*)malloc((2*map->capacity)*sizeof(opened_dir_struct));
+	assert(new_opened_files!=NULL);
+	map->capacity=2*map->capacity;
+
+		for(i=0;i<map->length;i++){
+			new_opened_files[i]=map->opened_files[i];
+
+		}
+		free(map->opened_files);
+		map->opened_files=new_opened_files;
+		return map;
+
+}
+//check if file path is a file or a directory
+int pathCheck(char *path){
+	int i;
+	struct stat statbuf;
+
+		if(stat(path,&statbuf)==0){
+			if (S_ISREG (statbuf.st_mode)){
+				i=0;
+			}
+			 if (S_ISDIR (statbuf.st_mode)) {
+			        i=-1;
+			    }
+
+		}
+		return i;
+}
+/* Opens a directory and store both the directoryfd and
+   the directory path in a structure
+*/
+
+opened_dir_struct * open_directory(char* file_path,opened_dir_struct *dos){
+	int dir_fd,k,j; char * dirname;
+	DIR *dir;
+	k=pathCheck(file_path);
+
+	if(k==0){
+		perror("");
+	}
+	else{
+		dirname=file_path;
+	}
+	dir=opendir(dirname);
+
+	if(dir !=NULL){
+		dir_fd=dirfd(dir);
+		dos->dirfd=dir_fd;
+		dos->dirname=dirname;
+
+	}
+	else{
+		perror("");
+	}
+
+	return dos;
+
+}
+
+
+Map* add_Opened_dirpath_map(opened_dir_struct ods){
+	int current=map->length,i;
+	map->opened_files[map->length]=ods;
+	map->length++;
+	return map;
+}
+Map* preopen(char* file,int mode){
+	int k;
+	char* dirname;
+	opened_dir_struct ods;
+	opened_dir_struct * odsp;
+	k=checkCapacity();
+	if(k<0){
+		map=increaseMapCapacity();
+	}
+
+	odsp=open_directory( file,&ods);
+	map=add_Opened_dirpath_map(*odsp);
+	return map;
+}
+
+/*
+  Finds how many characters in a string is in another
+  string begining from the first character
+*/
+int findMatchingChars(char *A,char *B){
+	//argu *A is path to be matched
+	//argu *B is a preopened dir path in the map pointer
+	int lenA,lenB,matchedNum=0,i;
+	assert(A!=NULL && B!=NULL);
```

Comment:
This is a fine assertion to make, but we need to explicitly declare the pre-condition in the header file.